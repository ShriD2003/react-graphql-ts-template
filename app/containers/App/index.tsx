/**
 *
 * App.js
 *
 * This component is the skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar)
 *
 */
import React, { useState, useCallback, useEffect } from 'react';
import GlobalStyle from '@app/global-styles';
import { routeConfig } from '@app/routeConfig';
import { Layout, Drawer, DrawerProps } from 'antd';
import map from 'lodash-es/map';
import { withRouter } from 'react-router';
import { Link, Route, Switch } from 'react-router-dom';
import { compose } from 'redux';
import styled, { ThemeProvider } from 'styled-components';
import For from '@components/For';
import Header from '@components/Header';
import { colors } from '@themes/index';
import icon from '@images/ion_rocket-sharp.svg';
import { CloseOutlined } from '@ant-design/icons';

const theme = {
  fg: colors.primary,
  bg: colors.secondaryText
};

const SidebarDrawer = styled(Drawer)`
  && {
    .ant-drawer-body {
      padding: 7rem 0 0 0;
      background-color: ${colors.primary};
      width: 6%;
      min-width: 4.5rem;
      max-width: 7rem;
      text-align: center;
    }
    .ant-drawer-close {
      top: 1rem;
    }
  }
`;

const SideBar = styled.div`
  && {
    width: 6%;
    min-width: 4.5rem;
    max-width: 7rem;
    min-height: calc(100vh - 7rem);
    height: auto;
    background-color: ${colors.primary};
    display: inline;
    text-align: center;
  }
`;

const RocketLogo = styled.img`
  && {
    margin-top: 1rem;
    object-fit: contain;
  }
`;

const CustomLayout = styled(Layout)`
  && {
    flex-direction: row;
  }
`;

export function App() {
  const [visible, setVisible] = useState(false);
  const [mobile, setMobile] = useState(window.innerWidth < 400);

  function detectMobile() {
    if (window.innerWidth < 400) {
      setMobile(true);
    } else {
      setMobile(false);
    }
  }

  useEffect(() => {
    window.addEventListener('resize', detectMobile);
    return () => {
      window.removeEventListener('resize', detectMobile);
    };
  }, []);

  const toggleSidebar = useCallback(() => setVisible((v) => !v), []);

  const sidebarProps: DrawerProps = mobile
    ? {
        closeIcon: <CloseOutlined style={{ color: colors.secondary, fontSize: '1.9rem' }} />,
        placement: 'left',
        visible,
        closable: true,
        onClose: toggleSidebar,
        width: 'max-content'
      }
    : {};

  const SidebarComponent = (props: DrawerProps) =>
    mobile ? <SidebarDrawer {...props} /> : <SideBar data-testid="sidebar" {...(props as any)} />;

  return (
    <ThemeProvider theme={theme}>
      <Header mobile={mobile} toggleSidebar={toggleSidebar} />
      <CustomLayout>
        <SidebarComponent {...sidebarProps}>
          <Link data-testid="rocket-home-link" aria-label="home link" to="/">
            <RocketLogo src={icon} alt="rocket-icon" />
          </Link>
        </SidebarComponent>
        <Layout.Content>
          <For
            ParentComponent={(props) => <Switch {...props} />}
            of={map(Object.keys(routeConfig))}
            renderItem={(routeKey, index) => {
              const Component = routeConfig[routeKey].component;
              return (
                <Route
                  exact={routeConfig[routeKey].exact}
                  key={index}
                  path={routeConfig[routeKey].route}
                  render={(props) => {
                    const updatedProps = {
                      ...props,
                      ...routeConfig[routeKey].props
                    };
                    return <Component {...updatedProps} />;
                  }}
                />
              );
            }}
          />
          <GlobalStyle />
        </Layout.Content>
      </CustomLayout>
    </ThemeProvider>
  );
}

export default compose(withRouter)(App);
