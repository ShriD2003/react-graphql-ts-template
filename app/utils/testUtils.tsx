import React from 'react';
import { I18nProvider } from '@lingui/react';
import { i18n } from '@lingui/core';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { Route, Router } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import configureStore from '@app/configureStore';
import { DEFAULT_LOCALE, translationMessages } from '@app/i18n';
import ConnectedLanguageProvider from '@containers/LanguageProvider';
import history from './history';

export const renderWithIntl = (children: React.ReactNode) => {
  i18n.load(DEFAULT_LOCALE, translationMessages[DEFAULT_LOCALE]);
  i18n.activate(DEFAULT_LOCALE);

  return render(<I18nProvider i18n={i18n}> {children} </I18nProvider>);
};

export const getComponentStyles = (Component: React.FC<any>, props = {}) => {
  renderWithIntl(Component(props));
  const { styledComponentId } = Component(props)!.type;
  const componentRoots = document.getElementsByClassName(styledComponentId);
  return window.getComputedStyle(componentRoots[0]);
};

export const renderProvider = (children: React.ReactNode, { path }: { path?: string } = {}, renderFn = render) => {
  const store = configureStore({}).store;
  return renderFn(
    <Provider store={store}>
      <ConnectedLanguageProvider messages={translationMessages}>
        <ThemeProvider
          theme={{
            main: 'violet'
          }}
        >
          <Router history={history}>{path ? <Route path={path}>{children}</Route> : children}</Router>
        </ThemeProvider>
      </ConnectedLanguageProvider>
    </Provider>
  );
};
export const timeout = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
export const apiResponseGenerator = <Data,>(ok: boolean, data: Data, error?: object) => ({
  ok,
  data,
  error
});
