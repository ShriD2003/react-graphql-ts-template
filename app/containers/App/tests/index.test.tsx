import React from 'react';
import { renderProvider } from '@utils/testUtils';
import App, { MOBILE_BREAKPOINT } from '../index';
import { BrowserRouter } from 'react-router-dom';
import { fireEvent, waitFor } from '@testing-library/react';

describe('<App /> container tests', () => {
  it('should render and match the snapshot', () => {
    const { container } = renderProvider(
      <BrowserRouter>
        <App />
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();
  });

  it('should set mobile to true window width is less MOBILE_BREAKPOINT', () => {
    Object.defineProperty(window, 'innerWidth', { value: MOBILE_BREAKPOINT - 50 });
    const { getByTestId } = renderProvider(<App />);
    global.dispatchEvent(new Event('resize'));
    expect(getByTestId('menu-icon')).toBeInTheDocument();
  });

  it('should set mobile to false window width is greater than MOBILE_BREAKPOINT', () => {
    Object.defineProperty(window, 'innerWidth', { value: MOBILE_BREAKPOINT + 50 });
    const { queryByTestId } = renderProvider(<App />);
    global.dispatchEvent(new Event('resize'));
    expect(queryByTestId('menu-icon')).not.toBeInTheDocument();
  });

  it('should open drawer in mobile if clicked on mobile icon', async () => {
    Object.defineProperty(window, 'innerWidth', { value: MOBILE_BREAKPOINT - 50 });
    const { getByTestId } = renderProvider(<App />);
    fireEvent.click(getByTestId('menu-icon'));
    await waitFor(() => expect(getByTestId('rocket-home-link')).toBeInTheDocument());
  });
});
