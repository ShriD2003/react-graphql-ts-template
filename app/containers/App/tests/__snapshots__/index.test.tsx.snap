// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> container tests should render and match the snapshot 1`] = `
<div>
  <header
    class="ant-layout-header Header__StyledHeader-a7tfu0-0 fquOgr"
    data-testid="header"
  >
    <img
      alt="logo"
      class="Header__Logo-a7tfu0-1 nUlCS"
      src="IMAGE_MOCK"
    />
    <p
      class="T__StyledText-gcrjip-0 fPfjuE Header__Title-a7tfu0-2 fdMcFf"
      data-testid="t"
    >
      Wednesday Solutions
    </p>
  </header>
  <main
    class="ant-layout-content"
  />
</div>
`;
